scripting/erm/ERMParser.cpp

```diff
@@ -346,27 +354,32 @@ namespace ERM
 			commentLine %= (~qi::char_("!") >> comment | (qi::char_('!') >> (~qi::char_("?!$#[")) >> comment ));

```

Comment:
Looks like this one can be a good candidate for unit tests. And I have seen some. Do they work?

Reply:
All existing tests are "green", but much more tests to be written

Reply:
Not for me so. They all fail with some internal error. Probably they are not able to load some lua related libs or something. Btw would not it be better to make them more modular? Test grammar separately and interpreter separately. Sure integrated test covers more and can save more time but it will be harder to debug them.

Reply:
I want to help you with unit tests related to VERM parser. I feel like it is the best way to get acquainted with it. But I want them being more granular. To accomplish this I either need static linkage or adding some extra DLL_EXPORT. What do you think about all this?

Something like this for example
`
TEST(ErmParser, SetVariable)
{
	std::stringstream source;
	source << "ZVSE" << std::endl;
	source << "!#VRv2:S0;" << std::endl;

	ERMInterpreter i(logGlobal);

	auto lua = i.loadScript("", source.str());
	std::vector<std::string> strings;
	boost::split(strings, lua, boost::is_any_of("\r\n"));

	ASSERT_EQ(strings.size(), 9) << lua;
	EXPECT_EQ(strings[5], "v['1'] = 0");
}`

Reply:
I decided not to link tests with script libs. If you want such level if granularity, you could access script module as `VLC->scriptHandler->erm` and use `scripting::Module::compile`

Reply:
> 
> 
> They all fail with some internal error. 

Is there any more info on errors?

Reply:
Nothing useful so far. Structural exception was thrown to the method body. Access violation or something. Will try to debug and see what is going on. I expect that test lacks some libs. It does not work out of the box on windows because AI and scripting dlls are in subfolders. Will try to figure it out.

Reply:
Is it possible to include ERM scripts to a map in a timed event? Or is it possible only in mods?

Reply:
> 
> 
> Is it possible to include ERM scripts to a map in a timed event? Or is it possible only in mods?

Only in a mod for now.

Reply:
Changed StdInc.h to include <luajit/lua.hpp> and rebuilt it all and now I have this Module not found: scripts/lib/erm

Reply:
Is that file "scripts/lib/erm.lua" (and the rest of scripts dir) present in data directory of your build?

Reply:
I have only build/bin/relwithdebinfo/scriptiong/lua51.dll and the rest of dlls.
No lib folder at all

Reply:
Where are .json files in build?

Reply:
Found them, they were absent in by build dir. Now all tests are green

Reply:
Today decided to try writing some tests and found a question. I tried to test multiline set
!#VRz103:S^Some
multiline
string^;
I am not sure it correctly escapes new lines. Where can I put erm file to test scripts in live?

Reply:
Here's example of one erm-vcmi script mod architecture: https://github.com/vcmi-mods/peasant-taxes

Reply:
Thank you, will check it.

Reply:
I try to run this script. It is messages.verm inside mod.
```
ZVSE
!#IF:M^Test test x^;
!?PI;
!!IF:M^Test x 1^;
!?HL-1;
!!IF:M^Test x 1^;
```

I have this error in console Script 'test-erm:testErm' failed to run, error: '[string "scripts/lib/erm"]:226: attempt to call a nil value'

Reply:
The problem node is !?HL trigger.

Reply:
Yes, there is no !?HL trigger yet.

Reply:
?!PI/!!IF says we received InfoWindow for non our player and no message appears. !#IF also does not work. What trigger/receiver can I use to visualize any string? Probably change creature description?

Reply:
Smth like this should work:
```
!#TM1:S1/999/1/255
!?TM1;
!!IF:M^Test x 1^;
```

Reply:
It works. And it works the way I expected
```
!!IF:M^Test
x1^;
```
results in message box with Test\nx1. So new lines does not work correctly.